styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedStateManager.java

```diff
@@ -228,6 +229,36 @@ private void initialize(WorkflowInstance workflowInstance) {
     }
   }
 
+  private void updateResourceCounters(StorageTransaction tx, Event event,
+                                      RunState currentRunState, RunState nextRunState) {
+    // increment counters if event is dequeue
+    if (isDequeue(event, nextRunState) && nextRunState.data().resourceIds().isPresent()) {
+      for (String resource : nextRunState.data().resourceIds().get()) {
+        tx.updateCounter(shardedCounter, resource, 1);
+      }
+    }
+
+    if (isConsumingResources(currentRunState.state())
+        && isReleasingResources(nextRunState.state())
+        && nextRunState.data().resourceIds().isPresent()) {
```

Comment:
Best practice would be to assert that resourceIds().isPresent. (Log an error if it isn't; it's a condition that should hold by design and if it doesn't we are in trouble. If we just express it as a benign if, the maintainer of code will think that it can be absent and will eventually go down odd paths when troubleshooting...)

Reply:
Or, hm, do we encounter cases when it's absent, during replay of legacy events or something?

Reply:
I couldn't come to a conclusion when following the flow, but to be safe, we can add a logline now. After deployment we can observe to see if we see those loglines on any cases where the resources might be empty, and use that to figure out what we have missed. If we don't see it happening, it should be safe to assume that resources are present everywhere.